### PR TITLE
add result Routes

### LINKED ISSUE #7: Provide Motivator Result Routes
According to the [API Specification](https://kopfsachen-dev.github.io/api/), the backend motivator service should provide the functionality to add and delete a result to/from the current motivator of a user.

### PR DESCRIPTION
**Pull request by alexfritsch10**
Fixes #7 

As @theEpsilon already worked on the middleware of this service, which can be seen in this PR #9, I didnt implement this again.

### CODE REVIEW

File: src/index.ts [hunk 1]
```diff
@@ -1,21 +1,23 @@
 import express from 'express';
 import log from 'loglevel';
-import db from './db';
+
+import * as db from './db';
```
Comment (jgraeger) [inline src/index.ts:4]: Is there a specific reason why you changed the import to this? Having to prepend `default` to each db member, makes the code significantly less expressive

File: src/index.ts [hunk 2]
```diff
@@ -25,6 +27,41 @@ app.get('/', (req, res) => {
   res.send('Hello world');
 });
 
+// add result to a specified motivator
+app.post('/motivator/result/:motivatorId', (req, res) => {
```
Comment (jgraeger) [inline src/index.ts:31]:
```suggestion
app.post('/motivator/:motivatorId/result', (req, res) => {
```
In the spec it's the other way around.

File: src/index.ts [hunk 3]
```diff
@@ -25,6 +27,41 @@ app.get('/', (req, res) => {
...
+  if (!req.body) {
+    res.status(422).send('Request body undefined');
+    return;
+  }
+
+  let { timestamp } = req.body.timestamp;
+
+  if (!timestamp) {
+    timestamp = new Date();
+  }
```
Comment (jgraeger) [inline src/index.ts:43]: `timestamp` is stated as a required field in the the API spec, but I like using the current server time and will change the spec accordingly :)

File: src/index.ts [hunk 4]
```diff
@@ -25,6 +27,41 @@ app.get('/', (req, res) => {
...
+});
+
+// delete result from a specified motivator
+app.delete('/motivator/result/:motivatorId', (req, res) => {
+  log.info('Deleting a result from a motivator');
+
+  // TODO: Auth UserID Header
+  db.default.models.motivatorResult.destroy({
+    where: { motivator_id: req.params.motivatorId },
+  }).then(() => res.status(204).send()).catch((err) => res.status(404).send(err));
```
Comment (jgraeger) [inline src/index.ts:62]: This deletes **all** results of a motivator, not only the ones of the user!

Comment (jgraeger) [inline src/index.ts:62]:
```suggestion
    where: { motivator_id: req.params.motivatorId },
  }).then(() => res.status(204).end()).catch((err) => res.status(404).end());
```
No request body for `404`.

Using send() on a 204 error is actually [fine according to the way it gets handles by express](https://github.com/expressjs/express/blob/master/lib/response.js#L111),although is not compliant with the express API. Use `end()` to send an empty response.

File: src/index.ts [hunk 5]
```diff
@@ -25,6 +27,41 @@ app.get('/', (req, res) => {
...
+  }
+
+  // TODO: Auth UserID Header
+  db.default.models.motivatorResult.create({
+    status: req.body.status,
+    timestamp: { timestamp },
+    feedback: req.body.feedback,
+    user_id: req.get('X-User-ID'),
+    motivator_id: req.params.motivatorId,
+  }).then((result) => res.status(201).send(result)).catch((err) => res.status(404).send(err));
```
Comment (jgraeger) [inline src/index.ts:52]:
```suggestion
  }).then((result) => res.status(201).send(result)).catch((err) => res.status(404).end());
```
`404 No content` errors should not contain any request body according to spec. As a general remark, please keep in mind, that just dumping the Database error to the user could disclose sensitive information. Carefully check the documentation of the [underlying APIs used](https://sequelize.org/api/v6/identifiers.html#errors), if that could be the case.

File: src/index.ts [hunk 6]
```diff
@@ -25,6 +27,41 @@ app.get('/', (req, res) => {
...
+  if (!timestamp) {
+    timestamp = new Date();
+  }
+
+  // TODO: Auth UserID Header
+  db.default.models.motivatorResult.create({
+    status: req.body.status,
+    timestamp: { timestamp },
+    feedback: req.body.feedback,
+    user_id: req.get('X-User-ID'),
```
Comment (jgraeger) [inline src/index.ts:50]: Remember to use the middleware later. You mentionend this in your PR opening message, so it's fine :)

File: src/index.ts [hunk 7]
```diff
@@ -25,6 +27,41 @@ app.get('/', (req, res) => {
   res.send('Hello world');
 });
 
+// add result to a specified motivator
+app.post('/motivator/result/:motivatorId', (req, res) => {
+  log.info('Adding a result to a motivator');
+
+  if (!req.body) {
+    res.status(422).send('Request body undefined');
```
Comment (jgraeger) [inline src/index.ts:35]:
```suggestion
    res.status(422).send({ error: 'Request body undefined'});
```
API stick required `422 Unprocessable entity` errors to return an object with the string property `error` containing the message.

File: src/index.ts [hunk 8]
```diff
@@ -25,6 +27,41 @@ app.get('/', (req, res) => {
...
+    timestamp = new Date();
+  }
+
+  // TODO: Auth UserID Header
+  db.default.models.motivatorResult.create({
+    status: req.body.status,
+    timestamp: { timestamp },
+    feedback: req.body.feedback,
+    user_id: req.get('X-User-ID'),
+    motivator_id: req.params.motivatorId,
```
Comment (jgraeger) [inline src/index.ts:51]: This is not storing any user input submitted.

**Review by jgraeger — CHANGES_REQUESTED** [review]
- Please move you're code to a [seperate express router](https://expressjs.com/de/guide/routing.html#express-router), as it's done in #9. This keeps our `index.ts` file clean. Enough complex stuff will be going on there.
- Please add validation on the entity provided with the `POST` request, so according to the API spec.
- See further comments and suggestion.

If that helps, you can auto generate code from the API spec and use it as an inspiration using `openapi-generator` (Albatrouz is also using this tool). It's a java program and can be installed via homebrew, but there is also a npm package `@openapitools/openapi-generator-cli` available (Please don't add it to the package.json dependencies if going for the latter :D). It includes generators for javascript and typescript server code, that you can use to *get inspiration* for your implementation. I would suggest going for the [typescript generator](https://openapi-generator.tech/docs/generators/typescript/), as it's supported by the project, gives you a little bit more safety and e.g. better autocompletion in your editor (if typescript is supported, like it is natively in vscode).

**Comment by alexfritsch10** [comment]
I opened a new PR on the branch "feature/gloriousResultRoutes" more building on what was already provided by @theEpsilon.

**Pull request closed without merging**